### ISSUE TITLE
Rebin Error for Linear Binning

### ISSUE THREAD

**Issue by k8macarthur** [hyperspy/hyperspy#1923]
```---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-16-d31d93845a2c> in <module>()
----> 1 s.rebin(scale=(3,3,1))

c:\users\macark\documents\github\hyperspy\hyperspy\_signals\eds.py in rebin(self, new_shape, scale, crop, out)
    182             new_shape=new_shape,
    183             scale=scale,)
--> 184         m = super().rebin(new_shape=new_shape, scale=scale, crop=crop, out=out)
    185         m = out or m
    186         time_factor = np.prod([factors[axis.index_in_array]

c:\users\macark\documents\github\hyperspy\hyperspy\signal.py in rebin(self, new_shape, scale, crop, out)
   2368         s = out or self._deepcopy_with_new_data(None, copy_variance=True)
   2369         data = hyperspy.misc.array_tools.rebin(
-> 2370             self.data, scale=factors, crop=crop)
   2371 
   2372         if out:

c:\users\macark\documents\github\hyperspy\hyperspy\misc\array_tools.py in rebin(a, new_shape, scale, crop)
    154     if _requires_linear_rebin(arr=a, scale=scale):
    155         _logger.debug("Using linear_bin")
--> 156         return _linear_bin(a, scale, crop)
    157     else:
    158         _logger.debug("Using standard rebin with lazy support")

c:\users\macark\documents\github\hyperspy\hyperspy\misc\array_tools.py in _linear_bin(dat, scale, crop)
    295         result = np.zeros((dim,) + dat.shape[1:], dtype=dtype)
    296         # Carry out binning over axis[0]
--> 297         _linear_bin_loop(result=result, data=dat, scale=s)
    298         # Swap axis[0] back to the original axis location.
    299         result = result.swapaxes(0, axis)

C:\ProgramData\Anaconda3\lib\site-packages\numba\dispatcher.py in _compile_for_args(self, *args, **kws)
    328                                 for i, err in failed_args))
    329                 e.patch_message(msg)
--> 330             raise e
    331 
    332     def inspect_llvm(self, signature=None):

C:\ProgramData\Anaconda3\lib\site-packages\numba\dispatcher.py in _compile_for_args(self, *args, **kws)
    305                 argtypes.append(self.typeof_pyval(a))
    306         try:
--> 307             return self.compile(tuple(argtypes))
    308         except errors.TypingError as e:
    309             # Intercept typing error that may be due to an argument

C:\ProgramData\Anaconda3\lib\site-packages\numba\dispatcher.py in compile(self, sig)
    577 
    578                 self._cache_misses[sig] += 1
--> 579                 cres = self._compiler.compile(args, return_type)
    580                 self.add_overload(cres)
    581                 self._cache.save_overload(sig, cres)

C:\ProgramData\Anaconda3\lib\site-packages\numba\dispatcher.py in compile(self, args, return_type)
     78                                       impl,
     79                                       args=args, return_type=return_type,
---> 80                                       flags=flags, locals=self.locals)
     81         # Check typing error if object mode is used
     82         if cres.typing_error is not None and not flags.enable_pyobject:

C:\ProgramData\Anaconda3\lib\site-packages\numba\compiler.py in compile_extra(typingctx, targetctx, func, args, return_type, flags, locals, library)
    764     pipeline = Pipeline(typingctx, targetctx, library,
    765                         args, return_type, flags, locals)
--> 766     return pipeline.compile_extra(func)
    767 
    768 

C:\ProgramData\Anaconda3\lib\site-packages\numba\compiler.py in compile_extra(self, func)
    360         self.lifted = ()
    361         self.lifted_from = None
--> 362         return self._compile_bytecode()
    363 
    364     def compile_ir(self, func_ir, lifted=(), lifted_from=None):

C:\ProgramData\Anaconda3\lib\site-packages\numba\compiler.py in _compile_bytecode(self)
    723         """
    724         assert self.func_ir is None
--> 725         return self._compile_core()
    726 
    727     def _compile_ir(self):

C:\ProgramData\Anaconda3\lib\site-packages\numba\compiler.py in _compile_core(self)
    710 
    711         pm.finalize()
--> 712         res = pm.run(self.status)
    713         if res is not None:
    714             # Early pipeline completion

C:\ProgramData\Anaconda3\lib\site-packages\numba\compiler.py in run(self, status)
    246                     # No more fallback pipelines?
    247                     if is_final_pipeline:
--> 248                         raise patched_exception
    249                     # Go to next fallback pipeline
    250                     else:

C:\ProgramData\Anaconda3\lib\site-packages\numba\compiler.py in run(self, status)
    238                 try:
    239                     event(stage_name)
--> 240                     stage()
    241                 except _EarlyPipelineCompletion as e:
    242                     return e.result

C:\ProgramData\Anaconda3\lib\site-packages\numba\compiler.py in stage_nopython_frontend(self)
    452                 self.args,
    453                 self.return_type,
--> 454                 self.locals)
    455 
    456         with self.fallback_context('Function "%s" has invalid return type'

C:\ProgramData\Anaconda3\lib\site-packages\numba\compiler.py in type_inference_stage(typingctx, interp, args, return_type, locals)
    866 
    867         infer.build_constraint()
--> 868         infer.propagate()
    869         typemap, restype, calltypes = infer.unify()
    870 

C:\ProgramData\Anaconda3\lib\site-packages\numba\typeinfer.py in propagate(self, raise_errors)
    842         if errors:
    843             if raise_errors:
--> 844                 raise errors[0]
    845             else:
    846                 return errors

TypingError: Failed at nopython (nopython frontend)
Internal error at <numba.typeinfer.IntrinsicCallConstraint object at 0x000002004FD43438>:
--%<-----------------------------------------------------------------
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\errors.py", line 259, in new_error_context
    yield
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typeinfer.py", line 501, in __call__
    self.resolve(typeinfer, typeinfer.typevars, fnty=self.func)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typeinfer.py", line 441, in resolve
    sig = typeinfer.resolve_call(fnty, pos_args, kw_args, literals=literals)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typeinfer.py", line 1113, in resolve_call
    literals=literals)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typing\context.py", line 189, in resolve_function_type
    res = defn.apply(args, kws)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typing\templates.py", line 207, in apply
    sig = generic(args, kws)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typing\npydecl.py", line 235, in generic
    assert len(sig.args) == 3
AttributeError: 'NoneType' object has no attribute 'args'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typeinfer.py", line 137, in propagate
    constraint(typeinfer)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typeinfer.py", line 501, in __call__
    self.resolve(typeinfer, typeinfer.typevars, fnty=self.func)
  File "C:\ProgramData\Anaconda3\lib\contextlib.py", line 99, in __exit__
    self.gen.throw(type, value, traceback)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\errors.py", line 265, in new_error_context
    six.reraise(type(newerr), newerr, sys.exc_info()[2])
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\six.py", line 658, in reraise
    raise value.with_traceback(tb)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\errors.py", line 259, in new_error_context
    yield
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typeinfer.py", line 501, in __call__
    self.resolve(typeinfer, typeinfer.typevars, fnty=self.func)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typeinfer.py", line 441, in resolve
    sig = typeinfer.resolve_call(fnty, pos_args, kw_args, literals=literals)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typeinfer.py", line 1113, in resolve_call
    literals=literals)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typing\context.py", line 189, in resolve_function_type
    res = defn.apply(args, kws)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typing\templates.py", line 207, in apply
    sig = generic(args, kws)
  File "C:\ProgramData\Anaconda3\lib\site-packages\numba\typing\npydecl.py", line 235, in generic
    assert len(sig.args) == 3
numba.errors.InternalError: 'NoneType' object has no attribute 'args'
[1] During: typing of intrinsic-call at c:\users\macark\documents\github\hyperspy\hyperspy\misc\array_tools.py (208)
--%<-----------------------------------------------------------------

File "..\github\hyperspy\hyperspy\misc\array_tools.py", line 208
```

**Comment by ericpre** [comment]
@k8macarthur, is from a development branch? If so can you pull the last changes?

**Comment by k8macarthur** [comment]
I've also checked with v 1.3.1 as well. Getting the same error. I'm wondering if it's a numba thing?

**Comment by ericpre** [comment]
Yes, it could be actually, numba version may need to match specific version of numpy, but you are using anaconda, so it should care about this... before updating can you check what is your numpy and numba version using:
```
conda list numpy
conda list numba
```

**Comment by k8macarthur** [comment]
```
numpy                     1.14.0           py36h4a99626_1
numpydoc                  0.7.0            py36ha25429e_0
numba                     0.36.2          np114py36h12cb543_0
```

**Comment by ericpre** [comment]
I have tested different environment (linux and windows) and different version and I can't reproduce. Can you try to reproduce using a new clean anaconda environment?
If not uninstalling numba will works but it will be slower.

**Comment by k8macarthur** [comment]
@jat255 was able to reproduce it: 
![image](https://user-images.githubusercontent.com/15925705/39110936-59423c24-46d3-11e8-8149-57b524b5aa20.png)

**Comment by ericpre** [comment]
Could you provide a minimum example, which it is not working? I have tested with:
``` python
import hyperspy.api as hs
s = hs.signals.Signal1D(np.arange(1000).reshape(10, 10, 10))
s2 = s.rebin(scale=(3,3,1))
```

It is possible that the bug has been there before, but it was (silently?) falling back to "python mode" and therefore very slow but working. Now we force not to use no python mode and it may explain why you see this error now. This obviously encourages users to report the bug! :wink:

**Comment by k8macarthur** [comment]
```
import hyperspy.api as hs
import numpy as np
s = hs.signals.Signal1D(np.random.rand(20,20,10))
s2 = s.rebin((3,3,1))
```

**Comment by k8macarthur** [comment]
```
s2 = s.rebin(scale=(3,3,1))
```
works but
```
s2 = s.rebin(new_shape=(3,3,1)) 
```
does not.

**Comment by k8macarthur** [comment]
My best guess is that it is a number vs numpy interaction see:
https://github.com/numba/numba/issues/2194

**Comment by k8macarthur** [comment]
Although fresh installing hyperspy this morning seems to have fixed the problem... I'm so confused...

**Comment by ericpre** [comment]
Indeed with numba, we need to be careful with "casting type" properly, see for example #1821. The error message from numba are usually not very clear and can be misleading too.
I can't reproduce the error with the code you copied above.
@jat255, can you confirm that a clean install fixes the issue?

**Comment by k8macarthur** [comment]
So I think my original error on EDX data occurred because the dtype was unit8 and not float.

**Comment by ericpre** [comment]
Now I can reproduce and this is actually slightly more subtle!
``` python
import hyperspy.api as hs
import numpy as np

s = hs.signals.Signal1D(np.arange(1000).reshape(10, 10, 10))
s.change_dtype(np.uint8)
s2 = s.rebin(scale=(3.0, 3.0, 1.0)) # work but the results is wrong!
s2 = s.rebin(scale=(3, 3, 1)) # doesn't work

```

**Comment by k8macarthur** [comment]
Interesting!  So there's a problem when dtype is an integer?

**Comment by ericpre** [comment]
Fixed in #1965.
